engine/src/QuadTree.cpp

```diff
@@ -0,0 +1,233 @@
+        else if (contains(southWest, node)) southWest->insert(node);
+        else if (contains(southEast, node)) southEast->insert(node);
+        else nodes.push_back(node);
+    }
+    else {
+        nodes.push_back(node);
+    }
+}
+
+bool QuadTree::split() {
+    if (isSplit) return true;
+
+    if (level == maxLevel || nodes.size() < node_capacity) return false;
+
+    unsigned int subLevel = level + 1;
+
+    unsigned int parentX = shape.getPosition().x;
+    unsigned int parentY = shape.getPosition().y;
+    unsigned int subWidth = shape.getSize().x / 2;
+    unsigned int subHeight = shape.getSize().y / 2;
+
+    northWest = new QuadTree(parentX,
+                             parentY,
+                             subWidth,
+                             subHeight,
+                             node_capacity, subLevel);
+    northEast = new QuadTree(parentX + subWidth,
+                             parentY,
+                             subWidth,
+                             subHeight,
+                             node_capacity, subLevel);
+    southWest = new QuadTree(parentX,
+                             parentY + subHeight,
+                             subWidth,
+                             subHeight,
+                             node_capacity, subLevel);
+    southEast = new QuadTree(parentX + subWidth,
+                             parentY + subHeight,
+                             subWidth,
+                             subHeight,
+                             node_capacity, subLevel);
+
+    isSplit = true;
+
+    std::list<sf::Sprite *> old_nodes = nodes;
+    nodes.clear();
+    for (auto node: old_nodes) {
+        insert(node);
+    }
+    
+    return true;
+}
+
+
+unsigned int QuadTree::nodes_size() const {
+    if (isSplit) {
+        return nodes.size() + northEast->nodes_size() + northWest->nodes_size() + southEast->nodes_size() +
+               southWest->nodes_size();
+    }
+    else return nodes.size();
+}
+
+bool QuadTree::isSplitUseful() const {
+    return nodes_size() > node_capacity;
+}
+
+std::list<sf::Sprite *> QuadTree::getNodesAt(const int &x, const int &y) {
+    std::list<sf::Sprite *> nodesAt;
+    if (isSplit) {
+        if (northWest->shape.getGlobalBounds().contains(x, y)) nodesAt.merge(northWest->getNodesAt(x, y));
+        if (northEast->shape.getGlobalBounds().contains(x, y)) nodesAt.merge(northEast->getNodesAt(x, y));
+        if (southWest->shape.getGlobalBounds().contains(x, y)) nodesAt.merge(southWest->getNodesAt(x, y));
+        if (southEast->shape.getGlobalBounds().contains(x, y)) nodesAt.merge(southEast->getNodesAt(x, y));
+    }
+    for (auto node: nodes) {
+        if (node->getGlobalBounds().contains(x, y)) nodesAt.push_back(node);
+    }
+    return nodesAt;
+}
+
+std::list<sf::Sprite *> QuadTree::getNodesAt(const sf::Vector2f pos) {
+    return getNodesAt(pos.x, pos.y);
+}
+
+void QuadTree::draw(sf::RenderTarget &canvas) {
```

Comment:
Make your quad tree derive of [`sf::Drawable`](http://www.sfml-dev.org/documentation/2.0-fr/classsf_1_1Drawable.php) instead. That's the best practice to draw stuff on screen, even it is for debugging.